R/data.table.R

```diff
@@ -2548,18 +2548,54 @@ setnames = function(x,old,new,skip_absent=FALSE) {
   invisible(x)
 }
 
-setcolorder = function(x, neworder=key(x))
+# in anchor mode user should provide a length 1 named vector: anchor = c(before/after = anchor_col_name/index)
+setcolorder = function(x, neworder=key(x), anchor = NULL)
 {
   if (is.character(neworder) && anyDuplicated(names(x)))
     stop("x has some duplicated column name(s): ", paste(names(x)[duplicated(names(x))], collapse=","), ". Please remove or rename the duplicate(s) and try again.")
   # if (!is.data.table(x)) stop("x is not a data.table")
   neworder = colnamesInt(x, neworder, check_dups=FALSE)  # dups are now checked inside Csetcolorder below
-  if (length(neworder) != length(x)) {
-    #if shorter than length(x), pad by the missing
-    #  elements (checks below will catch other mistakes)
-    neworder = c(neworder, setdiff(seq_along(x), neworder))
+  # original mode without anchor
+  if (is.null(anchor)) {
+    if (length(neworder) != length(x)) {
+      #if shorter than length(x), pad by the missing
+      #  elements (checks below will catch other mistakes)
+      final_order = c(neworder, setdiff(seq_along(x), neworder))
+    } else {
+      final_order <- neworder
+    }
+  } else {
+    if (!(names(anchor) %in% c("before", "after"))) {
+      stop("anchor need to be named with 'before' or 'after'")
+    }
+    if (!(class(anchor) %in% c("character", "numeric", "integer"))) {
+      stop("anchor need to be either column name or index")
+    }
+    # always operating in index mode, consistent with original mode
+    oldorder <- seq_along(x)
+    if (class(anchor) == "character") {
+      anchor_location <- which(names(x) == anchor)
+    } else {
+      anchor_location <- which(oldorder == anchor)
```

Comment:
This seems to simplify to ```anchor_location = as.integer(anchor)```. Along with it, you could do error handling of ```if(anchor_location < 1L || anchor_location > length(x)) stop("...")``` instead of relying on an zero-length integer from ```which```.